README.md

```diff
@@ -333,25 +333,125 @@ _You can enable the following settings in Xcode by running [this script](resourc
+    }
   ```
 
+  </details>
+
+* <a id='infer-property-types'></a>(<a href='#infer-property-types'>link</a>) **Prefer letting the type of a variable or property be inferred from the right-hand-side value rather than writing the type explicitly on the left-hand side.** [![SwiftFormat: preferInferredTypes](https://img.shields.io/badge/SwiftFormat-preferInferredTypes-7B0051.svg)](https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#preferInferredTypes)
+
+  <details>
+
+  Prefer using inferred types when the right-hand-side value is a static member with a leading dot (e.g. an `init`, a `static` property / function, or an enum case). This applies to both local variables and property declarations:
+
   ```swift
-  enum Direction {
-    case left
-    case right
+  // WRONG
+  struct SolarSystemBuilder {
+    let sun: Star = .init(mass: 1.989e30)
+    let earth: Planet = .earth
+
+    func setUp() {
+      let galaxy: Galaxy = .andromeda
+      let system: SolarSystem = .init(sun, earth)
+      galaxy.add(system)
+    }
   }
+  
+  // RIGHT
+  struct SolarSystemBuilder {
+    let sun = Star(mass: 1.989e30)
+    let earth = Planet.earth
 
-  func someDirection() -> Direction {
-    // WRONG
-    return Direction.left
+    func setUp() {
+      let galaxy = Galaxy.andromeda
+      let system = SolarSystem(sun, earth)
+      galaxy.add(system)
+    }
+  }
+  ```
 
-    // RIGHT
-    return .left
+  Explicit types are still permitted in other cases:
+
+  ```swift
+  // RIGHT: There is no right-hand-side value, so an explicit type is required.
+  let sun: Star
+
+  // RIGHT: The right-hand-side is not a static member of the left-hand type.
+  let moon: PlantaryBody = earth.moon
+  let sunMass: Float = 1.989e30
+  let planets: [Planet] = []
+  let venusMoon: Moon? = nil
+  ```
+
+  There are some rare cases where the inferred type syntax has a different meaning than the explicit type syntax. In these cases, the explicit type syntax is still permitted:
+
+  ```swift
+  extension String {
+    static let earth = "Earth"
   }
+
+  // WRONG: fails with "error: type 'String?' has no member 'earth'"
+  let planetName = String?.earth
+
+  // RIGHT
+  let planetName: String? = .earth
+  ```
+
+  ```swift
+  struct SaturnOutline: ShapeStyle { ... }
+
+  extension ShapeStyle where Self == SaturnOutline {
+    static var saturnOutline: SaturnOutline { 
+      SaturnOutline() 
+    }
+  }
+
+  // WRONG: fails with "error: static member 'saturnOutline' cannot be used on protocol metatype '(any ShapeStyle).Type'"
+  let myShape2 = (any ShapeStyle).myShape
+
+  // RIGHT: If the property's type is an existential / protocol type, moving the type
+  // to the right-hand side will result in invalid code if the value is defined in an
+  // extension like `extension ShapeStyle where Self == SaturnOutline`.
+  // SwiftFormat autocorrect detects this case by checking for the existential `any` keyword.
+  let myShape1: any ShapeStyle = .saturnOutline
```

Comment:
I spent a bit of time thinking about this case in more detail, that `let myShape1: ShapeStyle = .saturnOutline` is allowed but `let myShape1: = ShapeStyle.saturnOutline` isn't. I couldn't understand why this wasn't allowed.

This syntax was introduced in [SE-0299](https://github.com/apple/swift-evolution/blob/main/proposals/0299-extend-generic-static-member-lookup.md). This [came up briefly](https://forums.swift.org/t/se-0299-extending-static-member-lookup-in-generic-contexts/43958/33) in the review thread, and Pavel mentioned that there was more discussion in the pitch thread. A found the most compelling discussion [here](https://forums.swift.org/t/proposal-static-member-lookup-on-protocol-metatypes/41946/48) in the pitch thread.

They intentionally don't frame this as adding a static member to the protocol type, but rather a shorthand syntax that lets you omit the name of the concrete type.

Basically, the thinking is that this code:

```swift
let myShape1: ShapeStyle = .saturnOutline
```

is actually implicit shorthand for _this code_:

```swift
let myShape1: ShapeStyle = SaturnOutline.saturnOutline
```

_not_ shorthand for this code:

```swift
let myShape1: ShapeStyle = ShapeStyle.saturnOutline
```

This is subtle but distinct, and is a reasonable explanation for why `ShapeStyle.saturnOutline` isn't allowed.

Reply:
This problem should also away _eventually_, once `any` is required in a future language mode (just [not Swift 6](https://github.com/airbnb/swift/pull/263#discussion_r1561395487)).